File: whiteboard/app/src/main/java/com/ditto/whiteboard/transport/TransportLifecycleCoordinator.kt
package com.ditto.whiteboard.transport

internal fun isNearbyLifecycleActive(
  started: Boolean,
  foreground: Boolean,
  nearbyNetworkingRunning: Boolean,
  localOnly: Boolean,
): Boolean = started && foreground && nearbyNetworkingRunning && !localOnly

/**
 * Thread-safe desired lifecycle state for independently dispatched permission and foreground work.
 *
 * A granted permission is not ready for Ditto sync until refreshPermissions has completed for the
 * latest request. Consumers always reconcile a current snapshot after taking their lifecycle lock.
 */
internal class TransportLifecycleCoordinator(
  permissionsGranted: Boolean,
  requestedForeground: Boolean,
) {
  data class Snapshot(
    val permissionsGranted: Boolean,
    val permissionsReadyForSync: Boolean,
    val requestedForeground: Boolean,
    val permissionGeneration: Long,
  ) {
    fun shouldRunNearby(started: Boolean, localOnly: Boolean): Boolean =
      started &&
        !localOnly &&
        permissionsGranted &&
        permissionsReadyForSync &&
        requestedForeground
  }

  private var granted = permissionsGranted
  private var ready = permissionsGranted
  private var foreground = requestedForeground
  private var generation = 0L

  @Synchronized
  fun requestPermissions(allGranted: Boolean): Long {
    granted = allGranted
    // A false→true request cannot enable sync until the SDK has refreshed its permission view.
    if (!allGranted || !ready) ready = false
    generation += 1
    return generation
  }

  @Synchronized
  fun markPermissionsRefreshed(requestGeneration: Long): Snapshot? {
    if (requestGeneration != generation) return null
    ready = granted
    return snapshotLocked()
  }

  @Synchronized
  fun requestForeground(isForeground: Boolean): Boolean {
    if (foreground == isForeground) return false
    foreground = isForeground
    return true
  }

  @Synchronized
  fun snapshot(): Snapshot = snapshotLocked()

  @Synchronized
  fun isCurrent(requestGeneration: Long): Boolean = requestGeneration == generation

  private fun snapshotLocked() = Snapshot(
    permissionsGranted = granted,
    permissionsReadyForSync = ready,
    requestedForeground = foreground,
    permissionGeneration = generation,
  )
}

File: whiteboard-apple/Packages/WhiteboardCore/Sources/WhiteboardCore/Transport/TransportLifecycleCoordinator.swift
import Foundation

public func isNearbyLifecycleActive(
  started: Bool,
  foreground: Bool,
  nearbyNetworkingRunning: Bool,
  localOnly: Bool
) -> Bool {
  started && foreground && nearbyNetworkingRunning && !localOnly
}

/// Thread-safe desired lifecycle state for independently dispatched permission and foreground work.
///
/// A granted permission is not ready for Ditto sync until `markPermissionsRefreshed` has completed
/// for the latest request. Consumers always reconcile a current snapshot after taking their
/// lifecycle lock.
public final class TransportLifecycleCoordinator: @unchecked Sendable {
  public struct Snapshot: Equatable, Sendable {
    public var permissionsGranted: Bool
    public var permissionsReadyForSync: Bool
    public var requestedForeground: Bool
    public var permissionGeneration: Int64

    public init(
      permissionsGranted: Bool,
      permissionsReadyForSync: Bool,
      requestedForeground: Bool,
      permissionGeneration: Int64
    ) {
      self.permissionsGranted = permissionsGranted
      self.permissionsReadyForSync = permissionsReadyForSync
      self.requestedForeground = requestedForeground
      self.permissionGeneration = permissionGeneration
    }

    public func shouldRunNearby(started: Bool, localOnly: Bool) -> Bool {
      started && !localOnly && permissionsGranted && permissionsReadyForSync && requestedForeground
    }
  }

  private let lock = NSLock()
  private var granted: Bool
  private var ready: Bool
  private var foreground: Bool
  private var generation: Int64 = 0

  public init(permissionsGranted: Bool, requestedForeground: Bool) {
    granted = permissionsGranted
    ready = permissionsGranted
    foreground = requestedForeground
  }

  @discardableResult
  public func requestPermissions(allGranted: Bool) -> Int64 {
    lock.withLock {
      granted = allGranted
      // A false→true request cannot enable sync until the SDK has refreshed its permission view.
      if !allGranted || !ready { ready = false }
      generation += 1
      return generation
    }
  }

  public func markPermissionsRefreshed(requestGeneration: Int64) -> Snapshot? {
    lock.withLock {
      guard requestGeneration == generation else { return nil }
      ready = granted
      return snapshotLocked()
    }
  }

  @discardableResult
  public func requestForeground(_ isForeground: Bool) -> Bool {
    lock.withLock {
      guard foreground != isForeground else { return false }
      foreground = isForeground
      return true
    }
  }

  public func snapshot() -> Snapshot {
    lock.withLock { snapshotLocked() }
  }

  public func isCurrent(requestGeneration: Int64) -> Bool {
    lock.withLock { requestGeneration == generation }
  }

  private func snapshotLocked() -> Snapshot {
    Snapshot(
      permissionsGranted: granted,
      permissionsReadyForSync: ready,
      requestedForeground: foreground,
      permissionGeneration: generation
    )
  }
}
